mmpose/models/detectors/mesh.py

```diff
@@ -0,0 +1,298 @@
+            create_transl=False)
+
+        joints_regressor = torch.tensor(
+            np.load(smpl['joints_regressor']), dtype=torch.float).unsqueeze(0)
+        self.register_buffer('joints_regressor', joints_regressor)
+
+        self.with_gan = disc is not None and loss_gan is not None
+        if self.with_gan:
+            self.disc = SMPLDiscriminator(**disc)
+            self.loss_gan = builder.build_loss(loss_gan)
+        self.disc_step_count = 0
+
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+        self.loss_mesh = builder.build_loss(loss_mesh)
+        self.init_weights(pretrained=pretrained)
+
+    def init_weights(self, pretrained=None):
+        """Weight initialization for model."""
+        self.backbone.init_weights(pretrained)
+        self.mesh_head.init_weights()
+        if self.with_gan:
+            self.disc.init_weights()
+
+    def train_step(self, data_batch, optimizer, **kwargs):
+        """Train step function.
+
+        In this function, the detector will finish the train step following
+        the pipeline:
+            1. get fake and real SMPL parameters
+            2. optimize discriminator (if have)
+            3. optimize generator
+
+        If `self.train_cfg.disc_step > 1`, the train step will contain multiple
+        iterations for optimizing discriminator with different input data and
+        only one iteration for optimizing generator after `disc_step`
+        iterations for discriminator.
+
+        Args:
+            data_batch (torch.Tensor): Batch of data as input.
+            optimizer (dict[torch.optim.Optimizer]): Dict with optimizers for
+                generator and discriminator (if have).
+
+        Returns:
+            outputs (dict): Dict with loss, information for logger,
+            the number of samples.
+        """
+
+        img = data_batch['img']
+        pred_smpl = self.generator(img)
+        pred_pose, pred_beta, pred_camera = pred_smpl
+
+        # optimize discriminator (if have)
+        if self.train_cfg['disc_step'] > 0 and self.with_gan:
+            set_requires_grad(self.disc, True)
+            fake_data = (pred_camera.detach(), pred_pose.detach(),
+                         pred_beta.detach())
+            mosh_theta = data_batch['mosh_theta']
+            real_data = (mosh_theta[:, :3], mosh_theta[:,
+                                                       3:75], mosh_theta[:,
+                                                                         75:])
+            fake_score = self.disc(fake_data)
+            real_score = self.disc(real_data)
+
+            disc_losses = {}
+            disc_losses['real_loss'] = self.loss_gan(
+                real_score, target_is_real=True, is_disc=True)
+            disc_losses['fake_loss'] = self.loss_gan(
+                fake_score, target_is_real=False, is_disc=True)
+            loss_disc, log_vars_d = self._parse_losses(disc_losses)
+
+            optimizer['disc'].zero_grad()
```

Comment:
note that generator uses the full name

Reply:
OK, I will also use the full name for discriminator.